### PR TITLE
(WIP) updates for c++17 - opened for discussion

### PR DESCRIPTION
**Pull request by d-a-v**
It is not intended to be merged as-is, but rather is opened for discussion about the incoming use of g++-9.3 from https://github.com/esp8266/Arduino/pull/6294.

*updated*:

~Still not fixed in this PR:~
```
arduino/libraries/ESP8266Audio/src/libtinysoundfont/tsf.h: In function 'void tsf_voice_envelope_nextsegment(tsf_voice_envelope*, int, float)':
arduino/libraries/ESP8266Audio/src/libtinysoundfont/tsf.h:893:4: warning: this statement may fall through [-Wimplicit-fallthrough=]
  893 |    if (e->samplesUntilNextSegment > 0)
      |    ^~
```
~from [there](https://github.com/earlephilhower/ESP8266Audio/blob/aba310d3a022eb3d6a1b5644b62e657f9b6333db/src/libtinysoundfont/tsf.h#L900).~  *update:* fix is [upstream](https://github.com/schellingb/TinySoundFont/blob/21c07c0bcd702e7adf3db349ac926914b98d17ce/tsf.h#L829-L839)

@earlephilhower you may fix all of these without merging this questionable PR.

~Please note that `.c` modified files were renamed to `.cpp` because it appeared to me that `[[fallthrough]];` is c++ only (unless mistaken).~
It appears that `/* fall through */` is sufficient to keep gcc9 (c and c++) quiet

### CODE REVIEW

**Comment by d-a-v** [comment]
@earlephilhower You might prefer to sync with upstream libs !

**Comment by earlephilhower** [comment]
@d-a-v yeah, I think I need to.  I see he's got 34 more commits since I diverged.  I should have used submodules here instead of copying.  Didn't grok them when I started this whole thing, darn it.

File: src/libflac/bitreader.c [hunk 1]
```diff
@@ -30,6 +30,14 @@
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+#ifndef REGISTER
```
Comment (earlephilhower) [inline src/libflac/bitreader.c:33]: `register` is a hint that's ignored most times, anyway, so I think I would just drop the whole modifier everywhere.  GCC is smart enough, and we have ~12 open registers on the 8266 to work with, anyway...

File: src/libmad/huffman.c [hunk 2]
```diff
@@ -51,7 +51,7 @@
 # endif
 
 static
-const union huffquad const hufftabA[] PROGMEM = {
```
Comment (earlephilhower) [inline src/libmad/huffman.c:54]: What's the problem GCC has with `static const`?  Seems like the PROGMEM would require it anyway, no?

Comment (d-a-v) [reply src/libmad/huffman.c:54]: Presence of two `const` is annoying gcc.

File: src/libflac/bitreader.c (hunk 1 above)
Comment (d-a-v) [reply src/libflac/bitreader.c]: I agree but I didn't try the effect on gcc-4.8.2 (and whatever version is used with esp32)

**Comment by earlephilhower** [comment]
@d-a-v. I don't seem to be able to update your PR to the latest push which has the latest TSF (which might cover most/all of the tsf.h changes).  Think you could merge master on your end and re-push to see what's left?

**Comment by d-a-v** [comment]
Merge went without conflict on my side.

**Comment by earlephilhower** [comment]
One more merge, please.  All the submodules should be at latest revs now.  Sorry!

**Comment by d-a-v** [comment]
No worries, there were conflicts this time (some fall through already brought in)

**Comment by earlephilhower** [comment]
These seemed harmless and I didn't want to have you need to do manual merge fixes.  Anything else, just throw in a new PR.  Thx!

**Pull request merged**